File: python/bunsen/codes/loinc.py
"""
Support for importing the LOINC Hierarchy into Bunsen.
"""

from bunsen.codes import Hierarchies

def with_loinc_hierarchy(sparkSession, hierarchies, loinc_hierarchy_path, loinc_version):
    """
    Returns a hierarchies instance that includes the LOINC hierarchy read
    from the given location.

    :param sparkSession: the spark session
    :param hierarchies: the :class:`bunsen.codes.Hierarchies` class
        to which the LOINC hierarchy should be added.
    :param loinc_hierarchy_path: the path of the LOINC hierarchy to load.
        This can be any path compatible with Hadoop's FileSystem API.
    :param loinc_version: the version of LOINC that is being loaded
    :return: a :class:`bunsen.codes.Hierarchies` with the added content.
    """
    loinc = sparkSession._jvm.com.cerner.bunsen.codes.systems.Loinc

    jhierarchies = loinc.withLoincHierarchy(sparkSession._jsparkSession,
                                           hierarchies._jhierarchies,
                                           loinc_hierarchy_path,
                                           loinc_version)

    return Hierarchies(sparkSession, jhierarchies)

File: bunsen-core/src/main/java/com/cerner/bunsen/codes/systems/Loinc.java
package com.cerner.bunsen.codes.systems;

import static org.apache.spark.sql.functions.col;
import static org.apache.spark.sql.functions.lit;

import com.cerner.bunsen.codes.Hierarchies;
import com.cerner.bunsen.codes.Hierarchies.HierarchicalElement;
import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

/**
 * Support for LOINC.
 */
public class Loinc {

  /**
   * Hierarchy URI used for LOINC is-a relationships.
   */
  public static final String LOINC_HIERARCHY_URI = Hierarchies.HIERARCHY_URI_PREFIX + "loinc";

  /**
   * LOINC code system URI.
   */
  public static final String LOINC_CODE_SYSTEM_URI = "http://loinc.org";

  /**
   * Reads the LOINC mutliaxial hierarchy file and converts it to a {@link HierarchicalElement}
   * dataset.
   *
   * @param spark the Spark session
   * @param loincHierarchyPath path to the multiaxial hierarchy CSV
   * @return a dataset of {@link HierarchicalElement} representing the hierarchical relationship.
   */
  public static Dataset<HierarchicalElement> readMultiaxialHierarchyFile(SparkSession spark,
      String loincHierarchyPath) {

    return spark.read()
        .option("header", true)
        .csv(loincHierarchyPath)
        .select(col("IMMEDIATE_PARENT"), col("CODE"))
        .where(col("IMMEDIATE_PARENT").isNotNull()
            .and(col("IMMEDIATE_PARENT").notEqual(lit(""))))
        .where(col("CODE").isNotNull()
            .and(col("CODE").notEqual(lit(""))))
        .map((MapFunction<Row, HierarchicalElement>) row -> {

          HierarchicalElement element = new HierarchicalElement();

          element.setAncestorSystem(LOINC_CODE_SYSTEM_URI);
          element.setAncestorValue(row.getString(0));

          element.setDescendantSystem(LOINC_CODE_SYSTEM_URI);
          element.setDescendantValue(row.getString(1));

          return element;
        }, Hierarchies.getHierarchicalElementEncoder());
  }

  /**
   * Returns a {@link Hierarchies} instance with the specified multiaxial hierarchy. This method
   * reads the LOINC multiaxial hierarchy file and converts it to a {@link HierarchicalElement}
   * dataset, and adds it to the given hierarchies.
   *
   * @param spark the Spark session
   * @param hierarchies a {@link Hierarchies} instance to which the hierarchy will be added
   * @param loincHierarchyPath path to the multiaxial hierarchy CSV
   * @param loincVersion the version of the LOINC hierarchy being read
   * @return a {@link Hierarchies} instance that includes the read hierarchy.
   */
  public static Hierarchies withLoincHierarchy(SparkSession spark,
      Hierarchies hierarchies,
      String loincHierarchyPath,
      String loincVersion) {

    Dataset<HierarchicalElement> elements = readMultiaxialHierarchyFile(spark, loincHierarchyPath);

    return hierarchies.withHierarchyElements(LOINC_HIERARCHY_URI, loincVersion, elements);
  }
}
